# **Ruby Nouri Kermani - 30261323**

**Part 1: Data Cleaning and Preprocessing**

1.1 Load and Inspect the Dataset

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt


energy_data = pd.read_csv('/content/Building_Energy_Benchmarking.csv')
# loading the dataset

display(energy_data)
print('\n')
display(energy_data.shape)
# diplaying dataset's shape, columns, and data types


Property Id                    Property Name                  Address 1  \
0       10176804  Acadia Aquatic & Fitness Centre       9009 Fairmount Dr SE   
1        6169481                       Ad Valorem              2924 11 ST NE   
2        6305956             Alberta Trade Centre               315 10 AV SE   
3        8854296                   Andrew Davison                133 6 AV SE   
4       21988603           Animal Services Centre        2201 PORTLAND ST SE   
..           ...                              ...                        ...   
489     21988624                  Senior Services               311 34 AV SE   
490     21988625                 Fire Station #25              4705 76 AV SE   
491     21988627                           EMS #6       1940 WESTMOUNT BV NW   
492     21988628                 Fire Station #24             2607 106 AV SW   
493     21988629                 Fire Station #36  10071 HIDDEN VALLEY DR NW   

        City Postal Code Province Primary Property Type - Self Selected  \
0    Calgary     T2H 0Z4  Alberta                  Heated Swimming Pool   
1    Calgary      t2e7l7  Alberta                                Office   
2    Calgary     T2G 0W2  Alberta                                Office   
3    Calgary      T2G0G2  Alberta                                Office   
4    Calgary     T2G 4M7  Alberta                                Office   
..       ...         ...      ...                                   ...   
489  Calgary     T2G 1T7  Alberta                                Office   
490  Calgary     T2C 2X1  Alberta                          Fire Station   
491  Calgary     T2N 3G8  Alberta               Other - Public Services   
492  Calgary     T2W 4H7  Alberta                          Fire Station   
493  Calgary     T3A 5G1  Alberta                          Fire Station   

     Number of Buildings  Year Built Property GFA - Self-Reported (m²)  ...  \
0                      1        2010                              1838  ...   
1                      1        1981                              7770  ...   
2                      1        1974                              6681  ...   
3                      1        1979                             17468  ...   
4                      1        2000                              2143  ...   
..                   ...         ...                               ...  ...   
489                    1        1960                               768  ...   
490                    1        1982                              1108  ...   
491                    1        1964                               425  ...   
492                    1        1978                              1108  ...   
493                    1        1988                               204  ...   

     Direct GHG Emissions (Metric Tons CO2e)  \
0                                        280   
1                                        243   
2                                          0   
3                                         26   
4                                        253   
..                                       ...   
489                                       50   
490                                       53   
491                                       40   
492                                       41   
493                                       13   

    Direct GHG Emissions Intensity (kgCO2e/m²)  \
0                                        152.4   
1                                         31.2   
2                                          0.0   
3                                          1.5   
4                                        118.1   
..                                         ...   
489                                       65.6   
490                                       48.2   
491                                       94.5   
492                                       36.8   
493                                       61.3   

    Electricity 

(494, 31)

In [2]:
display(energy_data.info())
print('\n')
display(energy_data.dtypes)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 494 entries, 0 to 493
Data columns (total 31 columns):
 #   Column                                                                 Non-Null Count  Dtype  
---  ------                                                                 --------------  -----  
 0   Property Id                                                            494 non-null    int64  
 1   Property Name                                                          494 non-null    object 
 2   Address 1                                                              494 non-null    object 
 3   City                                                                   494 non-null    object 
 4   Postal Code                                                            494 non-null    object 
 5   Province                                                               494 non-null    object 
 6   Primary Property Type - Self Selected                                  494 non-null    obj

None

Property Id                                                                int64
Property Name                                                             object
Address 1                                                                 object
City                                                                      object
Postal Code                                                               object
Province                                                                  object
Primary Property Type - Self Selected                                     object
Number of Buildings                                                        int64
Year Built                                                                 int64
Property GFA - Self-Reported (m²)                                         object
ENERGY STAR Score                                                        float64
Site Energy Use (GJ)                                                      object
Weather Normalized Site Energy Use (GJ)                                   object
Site EUI (GJ/m²)                                                         float64
Weather Normalized Site EUI (GJ/m²)                                      float64
Source Energy Use (GJ)                                                    object
Weather Normalized Source Energy Use (GJ)                                 object
Source EUI (GJ/m²)                                                       float64
Weather Normalized Source EUI (GJ/m²)                                    float64
Total GHG Emissions (Metric Tons CO2e)                                    object
Total GHG Emissions Intensity (kgCO2e/m²)                                float64
Direct GHG Emissions (Metric Tons CO2e)                                   object
Direct GHG Emissions Intensity (kgCO2e/m²)                               float64
Electricity Use - Grid Purchase (kWh)                                     object
Natural Gas Use (GJ)                                                      object
District Hot Water Use (GJ)                                               object
Electricity Use – Generated from Onsite Renewable Systems (kWh)          float64
Green Power - Onsite and Offsite (kWh)                                   float64
Avoided Emissions - Onsite and Offsite Green Power (Metric Tons CO2e)    float64
Year Ending                                                                int64
Unique ID                                                                 object
dtype: object

In [3]:
missing_values_engdata = energy_data.isnull().sum()
print(missing_values_engdata)
# finding the number of missing values in each column

Property Id                                                                0
Property Name                                                              0
Address 1                                                                  0
City                                                                       0
Postal Code                                                                0
Province                                                                   0
Primary Property Type - Self Selected                                      0
Number of Buildings                                                        0
Year Built                                                                 0
Property GFA - Self-Reported (m²)                                          0
ENERGY STAR Score                                                        329
Site Energy Use (GJ)                                                       0
Weather Normalized Site Energy Use (GJ)                                    0

1.2 Handling Missing Data

In [4]:
missing_values_percentage = (energy_data.isnull().sum() / len(energy_data)) * 100
print(missing_values_percentage)
print('\n')
# finding the percentage of columns

drop_columns = missing_values_percentage[missing_values_percentage > 40].index
energy_data_dropped = energy_data.drop(columns=drop_columns)
# dropping columns with more than 40% missing values

print(energy_data_dropped.isnull().mean()*100)
# just to make sure what columns with missing values are left


Property Id                                                               0.000000
Property Name                                                             0.000000
Address 1                                                                 0.000000
City                                                                      0.000000
Postal Code                                                               0.000000
Province                                                                  0.000000
Primary Property Type - Self Selected                                     0.000000
Number of Buildings                                                       0.000000
Year Built                                                                0.000000
Property GFA - Self-Reported (m²)                                         0.000000
ENERGY STAR Score                                                        66.599190
Site Energy Use (GJ)                                                      0.000000
Weat

In [5]:
# column = Weather Normalized Source Energy Use (GJ)
# type = object - categorical column

# column = Natural Gas Use (GJ)
# type = object - categorical column

# using mode / most frequece
energy_data_filled = energy_data_dropped.fillna({
    'Weather Normalized Source Energy Use (GJ)': energy_data_dropped['Weather Normalized Source Energy Use (GJ)'].mode(),
    'Natural Gas Use (GJ)': energy_data_dropped['Natural Gas Use (GJ)'].mode()
})
# filling the missing values of categorical columns with mode

from sklearn.impute import SimpleImputer

num_columns = energy_data_filled.select_dtypes(include=['float64', 'int64']).columns
imputer = SimpleImputer(strategy = 'median')
energy_data_filled[num_columns] = imputer.fit_transform(energy_data_filled[num_columns])
# filling the missing values of numerical columns with median

display(energy_data_filled)


Property Id                    Property Name                  Address 1  \
0     10176804.0  Acadia Aquatic & Fitness Centre       9009 Fairmount Dr SE   
1      6169481.0                       Ad Valorem              2924 11 ST NE   
2      6305956.0             Alberta Trade Centre               315 10 AV SE   
3      8854296.0                   Andrew Davison                133 6 AV SE   
4     21988603.0           Animal Services Centre        2201 PORTLAND ST SE   
..           ...                              ...                        ...   
489   21988624.0                  Senior Services               311 34 AV SE   
490   21988625.0                 Fire Station #25              4705 76 AV SE   
491   21988627.0                           EMS #6       1940 WESTMOUNT BV NW   
492   21988628.0                 Fire Station #24             2607 106 AV SW   
493   21988629.0                 Fire Station #36  10071 HIDDEN VALLEY DR NW   

        City Postal Code Province Primary Property Type - Self Selected  \
0    Calgary     T2H 0Z4  Alberta                  Heated Swimming Pool   
1    Calgary      t2e7l7  Alberta                                Office   
2    Calgary     T2G 0W2  Alberta                                Office   
3    Calgary      T2G0G2  Alberta                                Office   
4    Calgary     T2G 4M7  Alberta                                Office   
..       ...         ...      ...                                   ...   
489  Calgary     T2G 1T7  Alberta                                Office   
490  Calgary     T2C 2X1  Alberta                          Fire Station   
491  Calgary     T2N 3G8  Alberta               Other - Public Services   
492  Calgary     T2W 4H7  Alberta                          Fire Station   
493  Calgary     T3A 5G1  Alberta                          Fire Station   

     Number of Buildings  Year Built Property GFA - Self-Reported (m²)  ...  \
0                    1.0      2010.0                              1838  ...   
1                    1.0      1981.0                              7770  ...   
2                    1.0      1974.0                              6681  ...   
3                    1.0      1979.0                             17468  ...   
4                    1.0      2000.0                              2143  ...   
..                   ...         ...                               ...  ...   
489                  1.0      1960.0                               768  ...   
490                  1.0      1982.0                              1108  ...   
491                  1.0      1964.0                               425  ...   
492                  1.0      1978.0                              1108  ...   
493                  1.0      1988.0                               204  ...   

    Source EUI (GJ/m²) Weather Normalized Source EUI (GJ/m²)  \
0                 4.47                                  4.91   
1                 1.57                                  1.64   
2                 0.94                                  1.00   
3                 0.97                                  1.04   
4                 4.00                                  4.27   
..                 ...                                   ...   
489               1.57                                  1.64   
490               1.49                                  1.49   
491               2.27                                  2.37   
492               1.37                                  1.35   
493               2.05                                  2.02   

     Total GHG Emissions (Metric Tons CO2e)  \
0                                       509   
1                                       920   
2                                       499   
3                                      1293   
4                                       568   
..                                      ...   
489                                      66   
490                                     105   
49

1.3 Extracting and Cleaning Data Using Regex

In [6]:
# Using Regex
print(energy_data_filled.dtypes)
dt_energy = energy_data_filled
# rename the dataframe for easy access

import re

def extract_numeric(value):
  value = str(value)
  match = re.findall(r"[\d,]+\.?\d*", value)
  # using regex to identify objects that are numeric text-values
  if match:
      return float(match[0].replace(",", ""))
      # remove commas, convert to float64
  return None

# Apply function to all object columns
object_columns = ["Property GFA - Self-Reported (m²)", "Site Energy Use (GJ)", "Weather Normalized Site Energy Use (GJ)",
                  "Source Energy Use (GJ)", "Weather Normalized Source Energy Use (GJ)",
                  "Total GHG Emissions (Metric Tons CO2e)", "Direct GHG Emissions (Metric Tons CO2e)",
                  "Electricity Use - Grid Purchase (kWh)", "Natural Gas Use (GJ)", "Unique ID"]

dt_energy[object_columns] = dt_energy[object_columns].applymap(extract_numeric)

print(dt_energy.iloc[:, 9:])
print(dt_energy.dtypes)


Property Id                                   float64
Property Name                                  object
Address 1                                      object
City                                           object
Postal Code                                    object
Province                                       object
Primary Property Type - Self Selected          object
Number of Buildings                           float64
Year Built                                    float64
Property GFA - Self-Reported (m²)              object
Site Energy Use (GJ)                           object
Weather Normalized Site Energy Use (GJ)        object
Site EUI (GJ/m²)                              float64
Weather Normalized Site EUI (GJ/m²)           float64
Source Energy Use (GJ)                         object
Weather Normalized Source Energy Use (GJ)      object
Source EUI (GJ/m²)                            float64
Weather Normalized Source EUI (GJ/m²)         float64
Total GHG Emissions (Metric 

<ipython-input-6-0a0923356587>:23: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  dt_energy[object_columns] = dt_energy[object_columns].applymap(extract_numeric)


In [7]:
# Postal codes in A1A 1A1 format

def format_postal_codes(postal_code):
  postal_code = re.sub(r"[^A-Za-z0-9]", "", postal_code).upper()
  # using regex
  # remove anything other than alphabet and numbers + change them to upper case
  if len(postal_code) == 6:
    return f"{postal_code[:3]} {postal_code[3:]}"
    # adding the space after the first 3 characters and before the last 3 characters
  return postal_code

dt_energy["Postal Code"] = dt_energy["Postal Code"].apply(format_postal_codes)
print(dt_energy["Postal Code"])

# the challenging part was to add the space


0      T2H 0Z4
1      T2E 7L7
2      T2G 0W2
3      T2G 0G2
4      T2G 4M7
        ...   
489    T2G 1T7
490    T2C 2X1
491    T2N 3G8
492    T2W 4H7
493    T3A 5G1
Name: Postal Code, Length: 494, dtype: object


In [8]:
# Clean and Extract meaningful text

# removing # from property names
# correcting the abbreviations

import re

def property_cleaning(name):
  name = re.sub(r"#", "", name)
  return name.strip()

dt_energy["Property Name"] = dt_energy["Property Name"].apply(property_cleaning)


def address_cleaning(address):
  address = address.lower()
  address = address.title()
  #changing all addresses to lower case and starting with Capital letter
  address = re.sub(r"\bst\b", "Street", address, flags=re.I)
  # using flags=re.I to make the string case insensitive
  address = re.sub(r"\b(ave|av)\b", "Avenue", address, flags=re.I)
  address = re.sub(r"\brd\b", "Road", address, flags=re.I)
  address = re.sub(r"\btr\b", "Trail", address, flags=re.I)
  address = re.sub(r"\b(bv|blvd)\b", "Boulevard", address, flags=re.I)
  address = re.sub(r"\bwy\b", "Way", address, flags=re.I)
  address = re.sub(r"\bdr\b", "Drive", address, flags=re.I)
  address = re.sub(r"\bne\b", "NE", address, flags=re.I)
  address = re.sub(r"\bnw\b", "NW", address, flags=re.I)
  address = re.sub(r"\bse\b", "SE", address, flags=re.I)
  address = re.sub(r"\bsw\b", "SW", address, flags=re.I)
  return address

dt_energy["Address 1"] = dt_energy["Address 1"].apply(address_cleaning)

print(dt_energy[["Property Name", "Address 1"]])


# https://docs.python.org/3/library/re.html

                       Property Name                     Address 1
0    Acadia Aquatic & Fitness Centre       9009 Fairmount Drive SE
1                         Ad Valorem             2924 11 Street NE
2               Alberta Trade Centre              315 10 Avenue SE
3                     Andrew Davison               133 6 Avenue SE
4             Animal Services Centre       2201 Portland Street SE
..                               ...                           ...
489                  Senior Services              311 34 Avenue SE
490                  Fire Station 25             4705 76 Avenue SE
491                            EMS 6   1940 Westmount Boulevard NW
492                  Fire Station 24            2607 106 Avenue SW
493                  Fire Station 36  10071 Hidden Valley Drive NW

[494 rows x 2 columns]


In [9]:
# Checking the numerical types

from pandas.api.types import is_numeric_dtype

columns_to_check = dt_energy.iloc[:, 9:]

for column in columns_to_check:
  column_dtype = dt_energy[column].dtype
  is_numeric = is_numeric_dtype(column_dtype)
  print(f"Is '{column}' numeric? {is_numeric}")

# https://pandas.pydata.org/docs/dev/reference/api/pandas.api.types.is_numeric_dtype.html

Is 'Property GFA - Self-Reported (m²)' numeric? True
Is 'Site Energy Use (GJ)' numeric? True
Is 'Weather Normalized Site Energy Use (GJ)' numeric? True
Is 'Site EUI (GJ/m²)' numeric? True
Is 'Weather Normalized Site EUI (GJ/m²)' numeric? True
Is 'Source Energy Use (GJ)' numeric? True
Is 'Weather Normalized Source Energy Use (GJ)' numeric? True
Is 'Source EUI (GJ/m²)' numeric? True
Is 'Weather Normalized Source EUI (GJ/m²)' numeric? True
Is 'Total GHG Emissions (Metric Tons CO2e)' numeric? True
Is 'Total GHG Emissions Intensity (kgCO2e/m²)' numeric? True
Is 'Direct GHG Emissions (Metric Tons CO2e)' numeric? True
Is 'Direct GHG Emissions Intensity (kgCO2e/m²)' numeric? True
Is 'Electricity Use - Grid Purchase (kWh)' numeric? True
Is 'Natural Gas Use (GJ)' numeric? True
Is 'Year Ending' numeric? True
Is 'Unique ID' numeric? True


**Part 2: Exploratory Data Analysis (EDA) and Aggregations**

2.1 Statistical Summary

In [10]:
# summary statistics for numerical features

numerical_features = dt_energy.select_dtypes(include=['float64', 'int64'])
summary_stats = numerical_features.describe()
display(summary_stats)

Property Id  Number of Buildings   Year Built  \
count  4.940000e+02           494.000000   494.000000   
mean   1.308877e+07             1.060729  1980.091093   
std    5.659556e+06             0.278281    25.159568   
min    6.169481e+06             1.000000  1896.000000   
25%    9.563763e+06             1.000000  1970.000000   
50%    9.997794e+06             1.000000  1978.000000   
75%    2.198860e+07             1.000000  1996.000000   
max    2.198863e+07             3.000000  2018.000000   

       Property GFA - Self-Reported (m²)  Site Energy Use (GJ)  \
count                         494.000000            494.000000   
mean                         4752.556883           8265.670040   
std                         10128.320688          19733.748811   
min                           113.000000            141.000000   
25%                          1096.650000           1232.825000   
50%                          1806.750000           2555.650000   
75%                          4257.900000           7999.050000   
max                         85941.000000         243202.000000   

       Weather Normalized Site Energy Use (GJ)  Site EUI (GJ/m²)  \
count                               494.000000        494.000000   
mean                               8397.193117          1.774798   
std                               19877.336939          1.306700   
min                                 146.000000          0.060000   
25%                                1256.150000          0.990000   
50%                                2572.500000          1.290000   
75%                                8006.575000          2.065000   
max                              242611.000000          8.430000   

       Weather Normalized Site EUI (GJ/m²)  Source Energy Use (GJ)  \
count                           494.000000              494.000000   
mean                              1.810324            10590.527733   
std                               1.331529            23438.866328   
min                               0.060000              237.000000   
25%                               1.000000             1601.525000   
50%                               1.310000             3238.600000   
75%                               2.110000            10242.400000   
max                               8.360000           261481.000000   

       Weather Normalized Source Energy Use (GJ)  Source EUI (GJ/m²)  \
count                                 492.000000          494.000000   
mean                                10250.205691            2.282753   
std                                 22908.704425            1.597846   
min                                   146.000000            0.110000   
25%                                  1561.450000            1.312500   
50%                                  3144.800000            1.680000   
75%                                  9647.925000            2.622500   
max                                260399.000000           10.130000   

       Weather Normalized Source EUI (GJ/m²)  \
count                             494.000000   
mean                                2.319717   
std                                 1.622897   
min                                 0.110000   
25%                                 1.350000   
50%                                 1.690000   
75%                                 2.652500   
max                                10.060000   

       Total GHG Emissions (Metric Tons CO2e)  \
count                              494.000000   
mean                               724.276316   
std                               1490.433401   
min                                 15.000000   
25%                                113.850000   
50%                                228.400000   
75%                                696.550000   
max                              13067.000000   

       Total GHG Emissions Intensity (kgCO2e/m²)  \
count                                 494.000000   
mean                       

In [11]:
# identify and explain key observations (outliers, mean vs. median difference)

# outliers:
def outlier_iqr(data, column):
  Q1 = data[column].quantile(0.25) # '25%' row in summary_stats
  Q3 = data[column].quantile(0.75) # '75%' row in summary_stats
  IQR = Q3 - Q1
  lower_bound = Q1 - 1.5 * IQR
  upper_bound = Q3 + 1.5 * IQR
  outliers = data[(data[column] < lower_bound) | (data[column] > upper_bound)]
  return outliers
# define a function to use IQR method

numerical_features = dt_energy.select_dtypes(include=['float64', 'int64'])
# recalling numerical columns

for column in numerical_features.columns:
  outliers = outlier_iqr(numerical_features, column)
  if not outliers.empty:
    print(f"Number of outliers in '{column}':")
    display(outliers.shape[0])
  else:
    print(f"No outliers found in '{column}'.")
# looping through numerical columns, one by one
# showing the number of outliers in each column



No outliers found in 'Property Id'.
Number of outliers in 'Number of Buildings':


25

Number of outliers in 'Year Built':


25

Number of outliers in 'Property GFA - Self-Reported (m²)':


60

Number of outliers in 'Site Energy Use (GJ)':


41

Number of outliers in 'Weather Normalized Site Energy Use (GJ)':


42

Number of outliers in 'Site EUI (GJ/m²)':


45

Number of outliers in 'Weather Normalized Site EUI (GJ/m²)':


44

Number of outliers in 'Source Energy Use (GJ)':


46

Number of outliers in 'Weather Normalized Source Energy Use (GJ)':


45

Number of outliers in 'Source EUI (GJ/m²)':


50

Number of outliers in 'Weather Normalized Source EUI (GJ/m²)':


51

Number of outliers in 'Total GHG Emissions (Metric Tons CO2e)':


46

Number of outliers in 'Total GHG Emissions Intensity (kgCO2e/m²)':


42

Number of outliers in 'Direct GHG Emissions (Metric Tons CO2e)':


37

Number of outliers in 'Direct GHG Emissions Intensity (kgCO2e/m²)':


55

Number of outliers in 'Electricity Use - Grid Purchase (kWh)':


51

Number of outliers in 'Natural Gas Use (GJ)':


35

No outliers found in 'Year Ending'.
No outliers found in 'Unique ID'.


In [12]:
# mean vs. median

def mean_vs_median(summary_stats):
  for column in summary_stats.columns:
    mean_value = summary_stats.loc['mean', column]
    # calling the 'mean' row from summary_stats
    median_value = summary_stats.loc['50%', column]
    # calling the '50%' or median from summary_stats
    print(f"Column: {column}")
    if mean_value > median_value:
      print("Right-skewed (mean > median)\n")
      # mean bigger than median - right-skewed distribution
    elif mean_value < median_value:
      print("Left-skewed (mean < median)\n")
      # mean smaller than median - left-skewed distribution
    else:
      print("Symmetrical (mean ~ median)\n")
      # approximately symmetrical (normal) distribution

# define a function to compare between Mean and Median

mean_vs_median(summary_stats)

# this comparison would have been better


Column: Property Id
Right-skewed (mean > median)

Column: Number of Buildings
Right-skewed (mean > median)

Column: Year Built
Right-skewed (mean > median)

Column: Property GFA - Self-Reported (m²)
Right-skewed (mean > median)

Column: Site Energy Use (GJ)
Right-skewed (mean > median)

Column: Weather Normalized Site Energy Use (GJ)
Right-skewed (mean > median)

Column: Site EUI (GJ/m²)
Right-skewed (mean > median)

Column: Weather Normalized Site EUI (GJ/m²)
Right-skewed (mean > median)

Column: Source Energy Use (GJ)
Right-skewed (mean > median)

Column: Weather Normalized Source Energy Use (GJ)
Right-skewed (mean > median)

Column: Source EUI (GJ/m²)
Right-skewed (mean > median)

Column: Weather Normalized Source EUI (GJ/m²)
Right-skewed (mean > median)

Column: Total GHG Emissions (Metric Tons CO2e)
Right-skewed (mean > median)

Column: Total GHG Emissions Intensity (kgCO2e/m²)
Right-skewed (mean > median)

Column: Direct GHG Emissions (Metric Tons CO2e)
Right-skewed (mean > media

2.2 Aggregations

In [13]:
# average Energy Use Intensity (EUI) by Property Type.

# not specified by Source EUI or Site EUI
# choosing columns with GJ/m² metric

# Site EUI (GJ/m²)
# Source EUI (GJ/m²)

eui_prop_type = dt_energy.groupby('Primary Property Type - Self Selected')[['Site EUI (GJ/m²)', 'Source EUI (GJ/m²)']].mean()
display(eui_prop_type)


Site EUI (GJ/m²)  \
Primary Property Type - Self Selected                                
Distribution Center                                       3.286000   
Fire Station                                              1.208827   
Fitness Center/Health Club/Gym                            4.385000   
Heated Swimming Pool                                      4.805333   
Ice/Curling Rink                                          2.182200   
Indoor Arena                                              1.106000   
Mixed Use Property                                        0.458000   
Museum                                                    1.584000   
Non-Refrigerated Warehouse                                0.768000   
Office                                                    1.519636   
Other                                                     0.070000   
Other - Public Services                                   1.526000   
Other - Recreation                                        2.165000   
Performing Arts                                           1.302000   
Repair Services (Vehicle, Shoe, Locksmith, etc.)          1.248000   
Self-Storage Facility                                     1.208000   
Social/Meeting Hall                                       1.550000   

                                                  Source EUI (GJ/m²)  
Primary Property Type - Self Selected                                 
Distribution Center                                         3.496000  
Fire Station                                                1.572849  
Fitness Center/Health Club/Gym                              5.306750  
Heated Swimming Pool                                        5.895333  
Ice/Curling Rink                                            2.862400  
Indoor Arena                                                1.358000  
Mixed Use Property                                          0.522000  
Museum                                                      2.050000  
Non-Refrigerated Warehouse                                  1.039333  
Office                                                      2.072545  
Other                                                       0.134000  
Other - Public Services                                     1.947600  
Other - Recreation                                          3.118000  
Performing Arts                                             1.694000  
Repair Services (Vehicle, Shoe, Locksmith, etc.)            1.525000  
Self-Storage Facility                                       1.562000  
Social/Meeting Hall                                         1.910000

In [14]:
# total Greenhouse Gas (GHG) emmisions by year

# not specified by year ending or year built.

# grouping by Year Built and Year Ending
ghg_year = dt_energy.groupby(['Year Built', 'Year Ending'])['Total GHG Emissions (Metric Tons CO2e)'].sum()
display(ghg_year)

Year Built  Year Ending
1896.0      2019.0            18.6
            2020.0            16.8
            2021.0            19.3
            2022.0            19.0
            2023.0            15.0
                            ...   
2018.0      2019.0          5148.8
            2020.0          6285.8
            2021.0          8222.9
            2022.0          9329.0
            2023.0         13475.0
Name: Total GHG Emissions (Metric Tons CO2e), Length: 255, dtype: float64

In [15]:
# top 5 properties with highest total energy consumption

# Energy Consumption includes:
  #Electricity Use - Grid Purchase (kWh)
  #Natural Gas Use (GJ)

# but Electricity use is in kWh and gas use is in GJ
# converting kwh to GJ


kwh_to_gj = 0.0036
# https://www.unitconverters.net/energy/kilowatt-hour-to-gigajoule.htm

dt_energy['Electricity Use (GJ)'] = dt_energy['Electricity Use - Grid Purchase (kWh)'] * kwh_to_gj
# creating a new column to put the converted values of electricity use

energy_by_prop = dt_energy.groupby('Primary Property Type - Self Selected')[['Electricity Use (GJ)', 'Natural Gas Use (GJ)']].sum()
# grouping and calculating the total
top_5 = energy_by_prop.nlargest(5, ['Electricity Use (GJ)', 'Natural Gas Use (GJ)'])
print(top_5)
# display top 5

print('\n')
#OR

# total energy = electricity use + gas use
dt_energy['Total Energy (GJ)'] = dt_energy['Electricity Use (GJ)'] + dt_energy['Natural Gas Use (GJ)']
# creating a new column for total energy
energy_by_prop1 = dt_energy.groupby('Primary Property Type - Self Selected')['Total Energy (GJ)'].sum()
# grouping and calculating the total
top_5 = energy_by_prop1.nlargest(5)
print(top_5)



                                       Electricity Use (GJ)  \
Primary Property Type - Self Selected                         
Office                                         441861.64488   
Ice/Curling Rink                               189620.12808   
Fitness Center/Health Club/Gym                 150872.17320   
Fire Station                                    95590.99404   
Non-Refrigerated Warehouse                      69139.28124   

                                       Natural Gas Use (GJ)  
Primary Property Type - Self Selected                        
Office                                             284387.7  
Ice/Curling Rink                                   378279.1  
Fitness Center/Health Club/Gym                     736042.0  
Fire Station                                       209388.1  
Non-Refrigerated Warehouse                          69318.1  


Primary Property Type - Self Selected
Fitness Center/Health Club/Gym    886914.17320
Distribution Center               726

2.3 Detecting Outliers Using Regex and IQR

In [16]:
# Using Regex only

# identify values that don't conform to expected numeric formats
import re

def find_numeric(value):
  if isinstance(value, (int, float)):
    return True
    # automatically matches the numeric format
  pattern = r"[^0-9.,\s]"
  # regex pattern to find any non-numeric values
  if re.search(pattern, str(value)):
    return False
    # it contains non-numeric
  else:
    return True
    # contains only numeric

is_it_numeric = dt_energy[:].map(find_numeric)
display(is_it_numeric)
# checking all rows and columns for numeric formats

Property Id  Property Name  Address 1   City  Postal Code  Province  \
0           True          False      False  False        False     False   
1           True          False      False  False        False     False   
2           True          False      False  False        False     False   
3           True          False      False  False        False     False   
4           True          False      False  False        False     False   
..           ...            ...        ...    ...          ...       ...   
489         True          False      False  False        False     False   
490         True          False      False  False        False     False   
491         True          False      False  False        False     False   
492         True          False      False  False        False     False   
493         True          False      False  False        False     False   

     Primary Property Type - Self Selected  Number of Buildings  Year Built  \
0                                    False                 True        True   
1                                    False                 True        True   
2                                    False                 True        True   
3                                    False                 True        True   
4                                    False                 True        True   
..                                     ...                  ...         ...   
489                                  False                 True        True   
490                                  False                 True        True   
491                                  False                 True        True   
492                                  False                 True        True   
493                                  False                 True        True   

     Property GFA - Self-Reported (m²)  ...  \
0                                 True  ...   
1                                 True  ...   
2                                 True  ...   
3                                 True  ...   
4                                 True  ...   
..                                 ...  ...   
489                               True  ...   
490                               True  ...   
491                               True  ...   
492                               True  ...   
493                               True  ...   

     Total GHG Emissions (Metric Tons CO2e)  \
0                                      True   
1                                      True   
2                                      True   
3                                      True   
4                                      True   
..                                      ...   
489                                    True   
490                                    True   
491                                    True   
492                                    True   
493                                    True   

     Total GHG Emissions Intensity (kgCO2e/m²)  \
0                                         True   
1                                         True   
2                                         True   
3                                         True   
4                                         True   
..                                         ...   
489                                       True   
490                                       True   
491                                       True   
492                                       True   
493                                       True   

     Direct GHG Emissions (Metric Tons CO2e)  \
0                                       True   
1                                       True   
2                                       True   
3                                       True   
4                                       True   
..                                       ...   
489                                     True   
490                         

In [17]:
# remove or correct incorrectly formatted numeric values.

import re

def clean_numeric(value):
  cleaned_value = re.sub(r"[^-\d.]", "", str(value))
  if cleaned_value.count(".") > 1:
    cleaned_value = cleaned_value.replace(".", "", cleaned_value.count(".") - 1)

  try:
    return float(cleaned_value)
  except ValueError:
    return value

display(dt_energy)

Property Id                    Property Name  \
0     10176804.0  Acadia Aquatic & Fitness Centre   
1      6169481.0                       Ad Valorem   
2      6305956.0             Alberta Trade Centre   
3      8854296.0                   Andrew Davison   
4     21988603.0           Animal Services Centre   
..           ...                              ...   
489   21988624.0                  Senior Services   
490   21988625.0                  Fire Station 25   
491   21988627.0                            EMS 6   
492   21988628.0                  Fire Station 24   
493   21988629.0                  Fire Station 36   

                        Address 1     City Postal Code Province  \
0         9009 Fairmount Drive SE  Calgary     T2H 0Z4  Alberta   
1               2924 11 Street NE  Calgary     T2E 7L7  Alberta   
2                315 10 Avenue SE  Calgary     T2G 0W2  Alberta   
3                 133 6 Avenue SE  Calgary     T2G 0G2  Alberta   
4         2201 Portland Street SE  Calgary     T2G 4M7  Alberta   
..                            ...      ...         ...      ...   
489              311 34 Avenue SE  Calgary     T2G 1T7  Alberta   
490             4705 76 Avenue SE  Calgary     T2C 2X1  Alberta   
491   1940 Westmount Boulevard NW  Calgary     T2N 3G8  Alberta   
492            2607 106 Avenue SW  Calgary     T2W 4H7  Alberta   
493  10071 Hidden Valley Drive NW  Calgary     T3A 5G1  Alberta   

    Primary Property Type - Self Selected  Number of Buildings  Year Built  \
0                    Heated Swimming Pool                  1.0      2010.0   
1                                  Office                  1.0      1981.0   
2                                  Office                  1.0      1974.0   
3                                  Office                  1.0      1979.0   
4                                  Office                  1.0      2000.0   
..                                    ...                  ...         ...   
489                                Office                  1.0      1960.0   
490                          Fire Station                  1.0      1982.0   
491               Other - Public Services                  1.0      1964.0   
492                          Fire Station                  1.0      1978.0   
493                          Fire Station                  1.0      1988.0   

     Property GFA - Self-Reported (m²)  ...  \
0                               1838.0  ...   
1                               7770.0  ...   
2                               6681.0  ...   
3                              17468.0  ...   
4                               2143.0  ...   
..                                 ...  ...   
489                              768.0  ...   
490                             1108.0  ...   
491                              425.0  ...   
492                             1108.0  ...   
493                              204.0  ...   

     Total GHG Emissions (Metric Tons CO2e)  \
0                                     509.0   
1                                     920.0   
2                                     499.0   
3                                    1293.0   
4                                     568.0   
..                                      ...   
489                                    66.0   
490                                   105.0   
491                                    53.0   
492                                   105.0   
493                                    28.0   

     Total GHG Emissions Intensity (kgCO2e/m²)  \
0                                        276.7   
1                                        118.4   
2                                         74.6   
3                                         74.0   
4                                        265.2   
..                                         ...   
489                                       86.3   
490                                       94.8   
491                                      124.8

**Part 3: Data Visualization**

3.1 Time-Series Visualization

3.2 Comparative Bar Charts

3.3 Heatmap Visualization

**Part 4: Further Analysis**

4.1 Correlation Analysis

4.2 Hypothesis Testing